core/app/c/[communitySlug]/stages/[stageId]/page.tsx

```diff
@@ -0,0 +1,80 @@
+import { Suspense } from "react";
+import Link from "next/link";
+import { notFound } from "next/navigation";
+
+import type { StagesId } from "db/public";
+import { Capabilities } from "db/src/public/Capabilities";
+import { MembershipType } from "db/src/public/MembershipType";
+import { Button } from "ui/button";
+
+import { CreatePubButton } from "~/app/components/pubs/CreatePubButton";
+import { getPageLoginData } from "~/lib/authentication/loginData";
+import { userCan } from "~/lib/authorization/capabilities";
+import { findCommunityBySlug } from "~/lib/server/community";
+import { getCommunityStages } from "~/lib/server/stages";
+import { PubListSkeleton } from "../../pubs/PubList";
+import { StagePubs } from "../components/StageList";
+
+export default async function Page({
+	params,
+	searchParams,
+}: {
+	searchParams: Record<string, string> & { page?: string };
+
+	params: { communitySlug: string; stageId: StagesId };
+}) {
+	const { communitySlug, stageId } = params;
+	const [{ user }, community] = await Promise.all([
+		getPageLoginData(),
+		findCommunityBySlug(communitySlug),
+	]);
+
+	if (!community) {
+		notFound();
+	}
+
+	const stage = await getCommunityStages({ stageId }).executeTakeFirstOrThrow();
+	const page = searchParams.page ? parseInt(searchParams.page) : 1;
+
+	const showEditButton = await userCan(
+		Capabilities.editCommunity,
+		{ type: MembershipType.community, communityId: community.id },
+		user.id
+	);
+
```

Comment:
Ideally, could you also add `title` metadata for this page? Together with `Promise.all`ing the two queries, it could look something like this

```suggestion
import type { Metadata } from "next";

import { cache, Suspense } from "react";
import Link from "next/link";
import { notFound } from "next/navigation";

import type { StagesId } from "db/public";
import { Capabilities } from "db/src/public/Capabilities";
import { MembershipType } from "db/src/public/MembershipType";
import { Button } from "ui/button";

import { CreatePubButton } from "~/app/components/pubs/CreatePubButton";
import { getPageLoginData } from "~/lib/authentication/loginData";
import { userCan } from "~/lib/authorization/capabilities";
import { findCommunityBySlug } from "~/lib/server/community";
import { getCommunityStages } from "~/lib/server/stages";
import { PubListSkeleton } from "../../pubs/PubList";
import { StagePubs } from "../components/StageList";

const getCommunityStagesCached = cache(async (stageId: StagesId) => {
	return getCommunityStages({ stageId }).executeTakeFirst();
});

export async function generateMetadata({
	params: { stageId },
}: {
	params: { stageId: string; communitySlug: string };
}): Promise<Metadata> {
	const stage = await getCommunityStagesCached(stageId as StagesId);
	if (!stage) {
		notFound();
	}

	return { title: `${stage.name} Stage` };
}

export default async function Page({
	params,
	searchParams,
}: {
	searchParams: Record<string, string> & { page?: string };

	params: { communitySlug: string; stageId: StagesId };
}) {
	const { communitySlug, stageId } = params;
	const [{ user }, community] = await Promise.all([
		getPageLoginData(),
		findCommunityBySlug(communitySlug),
	]);

	if (!community) {
		notFound();
	}

	const page = searchParams.page ? parseInt(searchParams.page) : 1;

	const stagePromise = getCommunityStagesCached(stageId);
	const showEditButtonPromise = userCan(
		Capabilities.editCommunity,
		{ type: MembershipType.community, communityId: community.id },
		user.id
	);

	const [stage, showEditButton] = await Promise.all([stagePromise, showEditButtonPromise]);

	if (!stage) {
		notFound();
	}

```

Reply:
moving the query outside of the page and wrapping it with `cache` is in order to have the query deduped in the metadata and body of the page.

Reply:
i'm not sure i follow the cache change. does autocache not work in this case for some reason?

Reply:
I've made the rest of the changes but didn't want to do this one without first understanding why!

Reply:
tldr: this optimization is not per se necessary, but i think it's probably better to do it than not.

ah sorry! this specifically adds the use of `React.cache`. this dedupes two function calls made in the same render pass, making that function only being called once during the render of the page (metadata + the body of the function). this is in addition to the `autoCache` behavior, which just stores the result of the query but also does some other stuff (calculating cache keys, reading from the cache) which does not get cached. using `React.cache` is  slightly more efficient (only slightly tho), because all the other stuff in `autoCache` will not happen.

basically, without adding react cache, this happens:
- in metadata: query gets compiled, cache keys calculated, query gets executed, result gets stored in filesystem cache 
- in body: query gets compiled, cache keys calculated, query gets reads from cache.

with react cache
- in metadata: query gets compiled, cache keys calculated, query gets executed, result gets stored in filesystem + in memory
- in body: result is immediately read from memory.

More info (not necessary):

i considered building this behavior into `autoCache`, but that would require changing the api somewhat. i discussed this with @3mcd at some point and he wasn't really a fan, so we decided to just do this kind of deduping manually when needed/desired.

for reference, the api for `autoCache` would need to look like this

```ts
const getPub = autoCache((pubId: PubsId) => db.selectFrom('pubs') )
```

rather than

```ts
const getPub =(pubId: PubsId) => autoCache(db.selectFrom('pubs'))
```


i.e. you would need to wrap the entire function inside of `autoCache`, rather than just the selection part. This is bc [calling different memoized fns will read from different caches](https://react.dev/reference/react/cache#pitfall-different-memoized-functions). Since currently calling eg `getCommunityStages(stageId)` will return a new object with new functions (`.execute`, `.executeTakeFirst`, `.executeTakeFirstOrThrow`), deduping is not possible.
 Plus, you would not be able to pass objects as arguments, as [react needs stable input references for the deduplication to work](https://react.dev/reference/react/cache#memoized-function-still-runs).

Maybe this is still somewhat unclear, in which case feel free to leave it out and merge it, and we can discuss this in more detail in some point if you're interested!